README.md
Description
===========

Installs and configures Visual Studio 2008 + SP1. 

This includes all features of Visual Studio 2008 including the 64-bit C++ compiler. The default Visual Studio 2008 installer does not enable the 64-bit C++ compiler installation. The "/full" option used here would enable 64-bit compile to install. No Documentation is installed. 

For some reason, the custom INI file for the silent install fails to launch on Windows 8.1/Server 2012 R2 due to dependencies. However, it launches fine if we don't use an INI file and use the "/full" option. :(

I added a registry key to disable the Windows compatibility popup for SQL 2005 install since it is not supported on Windows 8.1/Server 2012 R2. 

Our purpose was to run server configuration automation inside a corporate firewall with no internet connection for security reason. :( 

Please see usage below for more details.

This takes about 35 minutes to run on my machine since it downloads the ISO, unzips it to a local path and then runs the install. The ISOs are immediately deleted after a successful run. I tried running mount-diskimage for the ISO (available in Windows 8.1/2012 R2 or newer) to save time from the unzip proces. However, it kept crashing on dismount-diskimage of the ISO in order for me to delete it. :(


Requirements
============
* .Net Framework 3.5

Platform
--------

* Windows with Chef client 11.14.2

Tested on
---------

* Windows Server 2012 R2
* Windows 8.1

Cookbooks
---------

* Windows
* 7-zip
* ms_dotnet::ms_dotnet3 (.net 3.5 must be enabled on Windows 8.1/2012 R2)

Resources and Providers
=======================

<Any Resources Or Providers Here>

Attributes
==========

* node['vs-2008']['base-network-location']    = 'http://my-iis-redirect-server/visual-studio'
* node['vs-2008']['preserve_iso_file'] = true  #keeps the ISOs after installation. Default is false


Usage
=====

* Download Visual Studio 2008 Professional (x86) - ISO: from the source "https://msdn.microsoft.com/subscriptions/json/GetDownloadRequest?brand=MSDN&locale=en-US&fileId=14233&activexDisabled=true&akamaiDL=false")
* Download Visual Studio 2008 SP1 - ISO: https://msdn.microsoft.com/subscriptions/json/GetDownloadRequest?brand=MSDN&locale=en-us&fileId=37031&activexDisabled=true&akamaiDL=false
* Place the ISOs you downloaded on a network share 
* created and ran a Chef recipe (iis-packages-server) on a seperate machine to configure a Microsoft IIS that created an internal HTTP website to point the file share.
* Override the default attributes to match your network environment, ideally using a hostname instead of an IP (see attributes section, above)

The allowed us to serve the ISO, EXE, ZIP, MSI, etc... from an internal HTTP site. The HTTP is better than using Windows NTFS network file share since we don'tt have specify any user/password credentials.

Please view sample role file "sample.role.vs2008-professional-x14-26326.rb" for better understanding.


License and Author
==================

* Author:: Yeung Siu (yeung.siu@citrix.com)
* Contributors: Raymond Loiseau (ray6568@yahoo.com)
* Contributors: Adith Hegde (Adith.Hegde@citrix.com)

Copyright:: 2015 Citrix Systems, Inc.

Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at

    http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.


CHANGELOG.md
vs-2008 CHANGELOG
===================

This file is used to list changes made in each version of the vs-2008 cookbook.

1.0.9 
-----
- Yeung Siu - remove the .Net Framework 3.5 check for now since we need to do a delay check during execution time instead of compile time.

1.0.7
-----
- y.siu - Add a prerequisites check for .Net Framework 3.5 with a registry look up.

1.0.0
-----
- y.siu - Initial release of vs-2008

- - -


recipes/uninstall_sql.rb
#
# Author:: Joe Fitzgerald (<joe.fitzgerald@emc.com>)
# Cookbook Name:: vs-2008
# Recipe:: default
#
# Copyright 2012, EMC Corporation
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#     http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.
#


Chef::Log.info("try to uninstall sql 2005")

windows_package "Microsoft SQL Server 2005" do
  action :remove
end



libraries/prerequisites.rb
module VS2008
  module Prerequisites

    def is_vs2008_installed?()
      registry_key_exists?(node['vs-2008']['base-package-registry'])
    end

    def is_dotnet_installed?()
      #http://blogs.msdn.com/b/deva/archive/2008/12/10/how-to-determine-which-microsoft-net-framework-version-and-service-pack-installed.aspx
      registry_key_exists?(node['vs-2008']['dotnet_key_path'])      
    end
  
    def is_vs2008_sp1_installed?()
      registry_key_exists?(node['vs-2008']['sp1-package-registry'])
    end

    def is_vs2008_sp1_version_installed?()
      if (!is_vs2008_sp1_installed?())
        return false
      end

      registry_data_exists?(
        node['vs-2008']['sp1-package-registry'],
          { :name => node['vs-2008']['sp1-package-registry-name'], :type => :string, :data => node['vs-2008']['sp1-package-registry-value']}
      )     

    end
    def install_vs2008_sp1?()
      (node['vs-2008']['install-vs2008-sp1']).downcase.eql?('yes')
    end

  end

end


sample.role.vs2008-professional-x14-26326.rb
name "sample.role.vs2008-professional-x14-26326"
#This file format seems to be best decribed in 
# http://adamcod.es/2013/01/15/vagrant-is-easy-chef-is-hard-part2.html

# The "installers_path" needs to be defined in the solo.rb file or a file that it includes "Chef::Config.from_file "/tmp/vagrant-chef-3/custom-config.rb""
# In the Vagrantfile process, this is defined in the file listed under "Chef.custom_config_path" and right now it called "re-custom-windows"
location = 'http://my-iis-redirect-server/visual-studio'
base_iso_name = "en_visual_studio_2008_professional_x86_dvd_x14-26326.iso"
sp1_iso_name = "en_visual_studio_2008_service_pack_1_x86_dvd_x15-12962.iso"
default_attributes ({
    "vs-2008" => {
      "base-extraction-path" => File.join(Chef::Config[:file_cache_path], "vs2008"),
      "sp1-extraction-path" => File.join(Chef::Config[:file_cache_path], "vs2008sp1"),
      "base-network-location" => location,
      "base-iso-name" => base_iso_name,
      "base-iso-location" => File.join(location, base_iso_name),
      "base-iso-cache-location" => File.join(Chef::Config[:file_cache_path], base_iso_name).gsub("/", "\\"),
      "sp1-iso-location" => File.join(location, sp1_iso_name),
      "sp1-iso-name" => sp1_iso_name,
      "sp1-iso-cache-location" => File.join(Chef::Config[:file_cache_path], sp1_iso_name).gsub("/", "\\"),
      "sp1-installer" => 'vs90sp1\SPInstaller.exe',
      "base-iso-checksum" => "52EBF5731B75CCC460384CE3FD25BC984FB2D828AE51501EBAF0CADC27A33EE9",
      "sp1-iso-checksum" => "580F717269FAA10CF668140EF0A1A264CEC194E20A0083CB0D0004A897CC67",
      "base-display-name" => "Microsoft Visual Studio 2008 Professional Edition - ENU",
      "install_dir" => (ENV['ProgramFiles(x86)'] || 'C:\Program Files (x86)') + '\Microsoft Visual Studio 9.0',
      "devenv_exe_path" => '\Common7\IDE\devenv.exe',
      "install-options" =>  "/q /norestart /full /norestart",
      "sp1-install-options" => '/q /norestart',      
      "sp1-install-log" => File.join(Chef::Config[:file_cache_path], 'vs2008sp1\vs2008sp1.htm').gsub("/", "\\"),
      "timeout" => 6000

    }
  })



run_list(
  "recipe[vs-2008]",
  
)

 

recipes/default.rb
#
# Author:: Joe Fitzgerald (<joe.fitzgerald@emc.com>)
# Cookbook Name:: vs-2008
# Recipe:: default
#
# Copyright 2012, EMC Corporation
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#     http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.
#

::Chef::Recipe.send(:include, VS2008::Prerequisites)

#Verify if visual Studio  is installed
vs2008_is_installed = is_vs2008_installed?()

## TODO. Need to do a delay .Net Framework 3.5 check. On a fresh machine with no .Net 3.5, Chef evaluates this first before the "ms_dotnet::ms_dotnet3" is executed. Hence the result is false and "ms_dotnet::ms_dotnet3" could never start.
#ms_dotnet_35_is_installed = is_dotnet_installed?()
#Chef::Application.fatal!(".Net Framework 3.5 is not present") if (!ms_dotnet_35_is_installed)

Chef::Log.info("Starting Visual Studio 2008 installation.")
Chef::Log.info("Visual Studio 2008 is already installed!!! Skipping the installation.") if vs2008_is_installed

# Validate That ISOs Are Available In Specified Locations
# Unzip VS 2008 ISO To Staging Folder
# Ensure the installation ISO path has been set by the user other the install will not procede
if !node['vs-2008']['base-network-location'] 
  raise 'visual studio source attribute must be set before running this cookbook'
end


### Visual Studio 2008 install SQL Server 2005 which isn't support on Windows 8.1/2012 R2 or newer OS. This suppresses the popup and installs it.
registry_key node['vs-2008']['sql-server-appcompat']['path'] do
  values [{:name => node['vs-2008']['sql-server-appcompat']['key'] , :type => node['vs-2008']['sql-server-appcompat']['key-type'], :data => node['vs-2008']['sql-server-appcompat']['value'] }        
         ]
  action :create
end

vs2008_extraction_path = win_friendly_path(node['vs-2008']['base-extraction-path'])


# Extracting the ISO image
seven_zip_archive 'extracting_visual_studio_iso' do
  path vs2008_extraction_path
  source node['vs-2008']['base-iso-location']
  overwrite true
  checksum node['vs-2008']['base-iso-checksum']
  not_if { vs2008_is_installed }
end


# Install VS 2008
vs2008setup = File.join(node['vs-2008']['base-extraction-path'], 'setup\setup.exe').gsub("/", "\\")
Chef::Log.info("Installing Visual Studio 2008: #{vs2008setup}")  unless vs2008_is_installed
Chef::Log.info("install options ")  unless vs2008_is_installed

Chef::Log.info("vs2008_is_installed is #{vs2008_is_installed}")

windows_package "#{node['vs-2008']['base-display-name']}" do
  source "#{vs2008setup}"
  options node['vs-2008']['install-options']
  timeout node['vs-2008']['timeout']
  installer_type :custom
  notifies :delete, "file[#{node['vs-2008']['base-iso-cache-location']}]", :immediately
  notifies :delete, "directory[#{vs2008_extraction_path}]", :immediately
  not_if { vs2008_is_installed }
end


vs2008_sp1_is_installed = is_vs2008_sp1_version_installed?()
Chef::Log.info("Installing Visual Studio 2008 SP1.") 
Chef::Log.info("Visual Studio 2008 SP1 is already installed!!! Skipping the installation.") if vs2008_sp1_is_installed


vs2008_sp1_extraction_path = node['vs-2008']['sp1-extraction-path'].gsub("/", "\\")
vs2008_sp1_iso_location = node['vs-2008']['sp1-iso-location']
 
## Extracting the SP1 ISO image
seven_zip_archive 'extracting_visual_studio_sp1_iso' do
  path vs2008_sp1_extraction_path
  source vs2008_sp1_iso_location
  overwrite true
  checksum node['vs-2008']['sp1-iso-checksum']
  not_if {vs2008_sp1_is_installed}
 end


# Install VS 2008 SP1
vs2008sp1setup = File.join(node['vs-2008']['sp1-extraction-path'], node['vs-2008']['sp1-installer']).gsub("/", "\\")
vs2008sp1installopt = "#{node['vs-2008']['sp1-install-options']} /log #{node['vs-2008']['sp1-install-log']}"


Chef::Log.info("Running #{vs2008sp1setup} #{vs2008sp1installopt}")
windows_package "#{node['vs-2008']['sp1-display-name']}" do
  source "#{vs2008sp1setup}"
  options vs2008sp1installopt
  timeout node['vs-2008']['timeout']
  installer_type :custom
  action :install
  notifies :delete, "directory[#{vs2008_sp1_extraction_path}]", :immediately
  notifies :delete, "file[#{node['vs-2008']['sp1-iso-cache-location']}]", :immediately
  not_if { vs2008_sp1_is_installed }
#  notifies :request, 'windows_reboot[60]'
end

# Deleting the RTM ISO file from temp folder
file "#{node['vs-2008']['base-iso-cache-location']}" do
  action :nothing
  backup false
  not_if {node['vs-2008']['preserve_iso_file']}
end

# Deleting the SP1 ISO file from temp folder
file "#{node['vs-2008']['sp1-iso-cache-location']}" do
  action :nothing
  backup false
  not_if {node['vs-2008']['preserve_iso_file']}
end


# Cleanup extracted RTM files and directories
directory vs2008_extraction_path do
  action :nothing
  recursive true
  not_if {node['vs-2008']['preserve_extracted_file']}
end

# Cleanup extracted SP1 files and directories
directory vs2008_sp1_extraction_path do
  action :nothing
  recursive true
  not_if {node['vs-2008']['preserve_extracted_file']}
end
